subsys/bluetooth/controller/ll_sw/ull_conn_iso.c

```diff
@@ -496,53 +518,61 @@ static void cis_disabled_cb(void *param)
 	}
 }
 
-/**
- * @brief Stop and tear down a connected ISO stream
- * This function may be called to tear down a CIS. When the CIS teardown
- * has completed and the stream is released and callback is provided, the
- * cis_released_cb callback is invoked.
- *
- * @param cis		 Pointer to connected ISO stream to stop
- * @param cis_relased_cb Callback to invoke when the CIS has been released.
- *                       NULL to ignore.
- */
-void ull_conn_iso_cis_stop(struct ll_conn_iso_stream *cis,
-			   ll_iso_stream_released_cb_t cis_relased_cb)
+static void ticker_stop_op_cb(uint32_t status, void *param)
+{
+	static memq_link_t link;
+	static struct mayfly mfy = {0, 0, &link, NULL, cig_disable};
+	uint32_t ret;
+
+	/* Assert if race between thread and ULL */
+	LL_ASSERT(status == TICKER_STATUS_SUCCESS);
+
+	/* Check if any pending LLL events that need to be aborted */
+	mfy.param = param;
+	ret = mayfly_enqueue(TICKER_USER_ID_ULL_LOW,
+			     TICKER_USER_ID_ULL_HIGH, 0, &mfy);
+	LL_ASSERT(!ret);
+}
+
+static void cig_disable(void *param)
 {
 	struct ll_conn_iso_group *cig;
 	struct ull_hdr *hdr;
 
-	cig = cis->group;
+	/* Check ref count to determine if any pending LLL events in pipeline */
+	cig = param;
 	hdr = &cig->ull;
+	if (ull_ref_get(hdr)) {
+		static memq_link_t link;
+		static struct mayfly mfy = {0, 0, &link, NULL, lll_disable};
+		uint32_t ret;
 
-	if (cis->teardown) {
-		/* Teardown already started */
-		return;
-	}
-	cis->teardown = 1;
-	cis->released_cb = cis_relased_cb;
+		mfy.param = &cig->lll;
 
-	if (ull_ref_get(hdr)) {
-		/* Event is active (prepare/done ongoing) - wait for done and
-		 * continue CIS teardown from there. The disabled_cb cannot be
-		 * reserved for other use.
+		/* Setup disabled callback to be called when ref count
+		 * returns to zero.
 		 */
-		LL_ASSERT(!hdr->disabled_cb || hdr->disabled_cb == cis_disabled_cb);
+		LL_ASSERT(!hdr->disabled_cb);
+		hdr->disabled_param = mfy.param;
+		hdr->disabled_cb = cig_disabled_cb;
 
-		hdr->disabled_param = cig;
-		hdr->disabled_cb = cis_disabled_cb;
+		/* Trigger LLL disable */
+		ret = mayfly_enqueue(TICKER_USER_ID_ULL_HIGH,
+				     TICKER_USER_ID_LLL, 0, &mfy);
+		LL_ASSERT(!ret);
 	} else {
-		static memq_link_t link;
-		static struct mayfly mfy = {0, 0, &link, NULL, NULL};
-
-		/* Tear down CIS now in ULL_HIGH context. Ignore enqueue
-		 * error (already enqueued) as all CISes marked for teardown
-		 * will be handled in cis_disabled_cb. Use mayfly chaining to
-		 * prevent recursive stop calls.
-		 */
-		mfy.fp = cis_disabled_cb;
-		mfy.param = cig;
-		mayfly_enqueue(TICKER_USER_ID_ULL_LOW,
-			       TICKER_USER_ID_ULL_HIGH, 1, &mfy);
+		/* No pending LLL events */
+		cig_disabled_cb(&cig->lll);
```

Comment:
Calling cig_disabled_cb here directly in ULL_HIGH seems like the correct thing to do. However, this will cause recursing call to cis_disabled_cb, when more than one CIS is associated with an ACL connection - which it cannot handle. Either we need to make sure cis_released_cb is deferred, or for efficiency, we will need to guard for re-entrance into cis_disabled_cb to prevent stopping ticker twice.

Reply:
Give me a solution ;-), will chaining the call to cig_disabled_cb using mayfly help?